File: rust/collatz/MaxSequence.rs
const NUMBER: i32 = 500000;

fn collatz(mut x: i64) -> i64 {
    let mut len = 0;

    while x > 1 {
        if x % 2 == 0 {
            x = x / 2;
        } else {
            x = 3 * x + 1;
        }

        len += 1;
    }

    return len;
}

fn find_max_collatz(to: i64) -> (i64, i64) {
    let mut result: (i64, i64) = (1, 1);

    for number in 1..to {
        let len = collatz(number);

        if len > result.1 {
            result= (number, len);
        }
    }

    return result;
}

fn main() {
    let start = std::time::Instant::now();

    println!("{:?}", find_max_collatz(NUMBER as i64));

    let elapsed = start.elapsed();
    println!("Execution time: {}ms", elapsed.as_millis());
}

File: go/collatz/MaxSequence.go
package main

import (
	"fmt"
	"time"
)

const NUMBER = 500000

func collatz(x int) int {
	len := 0

	for x > 1 {
		if x%2 == 0 {
			x = x / 2
		} else {
			x = 3*x + 1
		}
		len++
	}

	return len
}

func findMaxCollatz(to int) (int, int) {
	result := [2]int{1, 1}

	for number := 1; number <= to; number++ {
		len := collatz(number)
		if len > result[1] {
			result = [2]int{number, len}
		}
	}

	return result[0], result[1]
}

func main() {
	startTime := time.Now()

	number, length := findMaxCollatz(NUMBER)
	fmt.Printf("[%d %d]\n", number, length)

	duration := time.Since(startTime)
	fmt.Printf("Execution time: %dms\n", duration.Milliseconds())
}
